### ISSUE TITLE
FP16 and FP32 weights question

### ISSUE THREAD

**Issue by InfiniteLife** [ultralytics/yolov5#1141]
## ❔Question
As for my understanding because of the line 
`    x['model'].half()  # to FP16`
in `utils.general` model trains in FP16  mode. And during training weights saved as FP32. And at the end of the training weights are saved as FP16.

I took intermediate weights(FP32) and converted them to ONNX, and from ONNX to Tensorflow. As I compared inference results of same model with Pytorch in FP16 vs same converted model Tensorflow FP32, I've noticed that Tensorflow accuracy is noticeably better.

1.  I wonder how is it possible since training allegedly happens in FP16? I understand model was saved as FP32 but from where comes that additional precision?

2. Where I need modify code to at the end of training save both variants of the network - in FP16 as it happens now, and in FP32 also?

**Comment by glenn-jocher** [comment]
@InfiniteLife fp16 and fp32 mAPs are identical in pytorch.

**Comment by yinyuhangi** [comment]
@InfiniteLife    I have the same problem with you. How can I save only the FP32 model after training?

**Comment by H19012** [comment]
> ## ❔Question
> As for my understanding because of the line
> ` x['model'].half() # to FP16`
> in `utils.general` model trains in FP16 mode. And during training weights saved as FP32. And at the end of the training weights are saved as FP16.
> 
> I took intermediate weights(FP32) and converted them to ONNX, and from ONNX to Tensorflow. As I compared inference results of same model with Pytorch in FP16 vs same converted model Tensorflow FP32, I've noticed that Tensorflow accuracy is noticeably better.
> 
> 1. I wonder how is it possible since training allegedly happens in FP16? I understand model was saved as FP32 but from where comes that additional precision?
> 2. Where I need modify code to at the end of training save both variants of the network - in FP16 as it happens now, and in FP32 also?
> 
> ## Additional context

How did you save the FP32 weights?
